lib/src/working_copy.rs

```diff
@@ -391,18 +396,92 @@ impl TreeState {
         Ok(self.store.write_symlink(path, str_target)?)
     }
 
+    #[tokio::main]
+    async fn query_fsmonitor(
+        &mut self,
+    ) -> Result<(FsmonitorClock, Option<Vec<PathBuf>>), FsmonitorError> {
+        let working_copy_path = self.working_copy_path.to_owned();
+        tokio::spawn(async move {
+            let fsmonitor = Fsmonitor::init(&working_copy_path).await?;
+            let previous_clock = None;
+            fsmonitor.query_changed_files(previous_clock).await
+        })
+        .await
+        .unwrap()
+    }
+
     /// Look for changes to the working copy. If there are any changes, create
     /// a new tree from it and return it, and also update the dirstate on disk.
-    pub fn snapshot(&mut self, base_ignores: Arc<GitIgnoreFile>) -> Result<TreeId, SnapshotError> {
+    pub fn snapshot(
+        &mut self,
+        base_ignores: Arc<GitIgnoreFile>,
+        should_use_fsmonitor: bool,
+    ) -> Result<TreeId, SnapshotError> {
+        let (_fsmonitor_clock, changed_files) = if should_use_fsmonitor {
+            match self.query_fsmonitor() {
+                Ok((fsmonitor_clock, changed_files)) => (Some(fsmonitor_clock), changed_files),
+                Err(err) => {
+                    eprintln!("Failed to query fsmonitor: {}", err);
+                    (None, None)
+                }
+            }
+        } else {
+            (None, None)
+        };
+
         let sparse_matcher = self.sparse_matcher();
-        let mut work = vec![(
-            RepoPath::root(),
-            self.working_copy_path.clone(),
-            base_ignores,
-        )];
         let mut tree_builder = self.store.tree_builder(self.tree_id.clone());
         let mut deleted_files: HashSet<_> = self.file_states.keys().cloned().collect();
-        while let Some((dir, disk_dir, git_ignore)) = work.pop() {
+
+        struct WorkItem {
+            dir: RepoPath,
+            disk_dir: PathBuf,
+            git_ignore: Arc<GitIgnoreFile>,
+            should_recurse: bool,
+        }
+        let repo_root = RepoPath::root();
+        let repo_root_pathbuf = repo_root.to_fs_path(&self.working_copy_path);
+        let mut work: Vec<WorkItem> = match changed_files {
+            Some(changed_files) => changed_files
+                .into_iter()
+                .map(|path| {
+                    let repo_path_components = path
+                        .iter()
+                        .map(|component| component.to_str().expect("Non-UTF-8 path component"))
+                        .map(RepoPathComponent::from)
+                        .collect();
+                    let dir = RepoPath::from_components(repo_path_components);
+                    WorkItem {
+                        dir,
+                        disk_dir: repo_root_pathbuf.join(path),
+
+                        // FIXME: it's possible that we missed a `.gitignore`
```

Comment:
I think we can fix this by creating a `FilesMatcher` from `changed_files` and then intersect that with the `sparse_matcher`. It's just that we don't have a simple way of intersecting matchers yet. However, I have one lying around from some unfinished work on making auto-add optional (#323). Let me send that and you can see if that helps. To clarify, the idea is to always create a single initial `WorkItem` (thanks for creating a struct for that, btw) and use different matchers instead.